src/ucp/core/ucp_am.c

```diff
@@ -558,189 +559,181 @@ ucp_am_find_first_rdesc(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
 {
     uint16_t am_id;
     ucs_status_t status;
-    ucp_am_first_hdr_t *f_hdr;
+    ucp_am_first_hdr_t *first_hdr;
+    void *msg;
     ucp_ep_h reply_ep;
 
-    ucs_assert(f_rdesc->am_first.data != NULL);
-
-    memcpy(UCS_PTR_BYTE_OFFSET(f_rdesc->am_first.data + 1, offset),
-           data, length);
+    ucp_am_copy_data_fragment(first_rdesc, data, length, offset);
 
-    f_rdesc->am_first.left -= length;
-    if (f_rdesc->am_first.left > 0) {
+    if (first_rdesc->am_first.remaining > 0) {
         /* not all fragments arrived yet */
-        return UCS_INPROGRESS;
+        return;
     }
 
-    f_hdr = (ucp_am_first_hdr_t*)(f_rdesc + 1);
-    am_id = f_hdr->super.super.am_id;
+    first_hdr = (ucp_am_first_hdr_t*)(first_rdesc + 1);
+    am_id     = first_hdr->super.super.am_id;
+    msg       = first_hdr + 1;
 
-    if (!ucp_am_recv_check_id(worker, am_id)) {
-        status = UCS_OK;
-    } else {
-        reply_ep = (f_hdr->super.super.flags & UCP_AM_SEND_REPLY)        ?
-                   ucp_worker_get_ep_by_ptr(worker, f_hdr->super.ep_ptr) :
+    if (ucs_likely(ucp_am_recv_check_id(worker, am_id))) {
+        reply_ep = (first_hdr->super.super.flags & UCP_AM_SEND_REPLY)        ?
+                   ucp_worker_get_ep_by_ptr(worker, first_hdr->super.ep_ptr) :
                    NULL;
-        status   = worker->am.cbs[am_id].cb(worker->am.cbs[am_id].context,
-                                            f_rdesc->am_first.data + 1,
-                                            f_hdr->total_size, reply_ep,
+        status   = worker->am.cbs[am_id].cb(worker->am.cbs[am_id].context, msg,
+                                            first_hdr->total_size, reply_ep,
                                             UCP_CB_PARAM_FLAG_DATA);
-    }
-
-    if (status != UCS_INPROGRESS) {
-        ucs_free(f_rdesc->am_first.data); /* user does not need to hold this data */
+    } else {
+        status = UCS_OK;
     }
 
     /* message assembled, remove first fragment descriptor from the list in
      * ep AM extension */
-    ucs_list_del(&f_rdesc->am_first.list);
-
-    return UCS_OK;
-}
+    ucs_list_del(&first_rdesc->am_first.list);
 
-static UCS_F_ALWAYS_INLINE int
-ucp_am_check_rdesc_msg_id(ucp_recv_desc_t *rdesc, uint64_t msg_id)
-{
-    ucp_am_mid_hdr_t *m_hdr = (ucp_am_mid_hdr_t*)(rdesc + 1);
+    if (status != UCS_INPROGRESS) {
+        /* user does not need to hold this data */
+        ucs_free(first_rdesc);
+        return;
+    }
 
-    return m_hdr->msg_id == msg_id;
+    /* Need to reinit descriptor, because we passed data shifted by
+     * ucp_am_first_hdr_t size to the cb. Also will need to step back by header
+     * size for freing malloced data in ucp_am_data_release (so save
+     * header size in uct_desc_offset for convinience). */
+    first_rdesc                  = (ucp_recv_desc_t*)msg - 1;
+    first_rdesc->flags           = UCP_RECV_DESC_FLAG_MALLOC;
+    first_rdesc->uct_desc_offset = sizeof(*first_hdr);
```

Comment:
why do we need to update ? isn't it going to be released anyway?

Reply:
we have to shift the rdesc by the first header size, because we provide pointer to the payload to the cb. So, when user will call am_data_release, it will calculate rdesc pointer incorrectly

Reply:
is this when we returning rdesc, or returning malloc'ed memory?
aren't we doing this already in line 680?

Reply:
we can remove `first_rdesc->flags           = UCP_RECV_DESC_FLAG_MALLOC;` on line 680. Left it for documentation purposes

Reply:
won't it be better to keep only the 680 one? then no need to calc first_rdesc here as well

Reply:
we have to do it here, because:
1) at first we allocate memory and layout is   | rdesc | first_hdr | data |
2) while we assemble the message we use first_hdr, getting it from rdesc
3) we pass pointer to the data to the user callback, so when he releases this data, we have to do `desc = data - sizeof(desc)`,
so we will not get an original desc. 
4) we reinitialize rdesc here by stepping forward on `first_hdr` size and init `flags` field
5) when data is released we get new rdesc initialized on step 4) check its flags and notice it was malloced. Then we step back by `first_hdr` size and get the original pointer to free it